planning/MartenV4.md

```diff
@@ -0,0 +1,114 @@
+# Marten v4 Ideas
+
+## Pulling it Off
+
+We've got the typical problems of needing to address incoming pull requests and bug issues in master while probably needing to have a long lived branch for *v4*. 
+
+Here are some thoughts:
+
+* First do a bug sweep 3.12 release to address as many of the tactical problems as we can before branching to v4?
+* Start with the unit testing improvements as a way to speed up the build before we dive into much more of this? 
+* Possibly do a series of v4, then v5 releases to do this in smaller chunks? We've mostly said do the event store as v4, then Linq improvements as v5
+* Do the Event Store v4 work in a separate project built as an add on from the very beginning, but leave the existing event store in place. That would enable us to do a lot of work *and* mostly be able to keep that work in master so we don't have long-lived branch problems
```

Comment:
That seems like an idea, but currently, it's hard for me to say if that would be the best option. It would depend on how revolutionary will be the changes. If they are the only revolution inside but not in the public API then probably we can even put it into the core part. We need to remember that keeping it separate code it's not much different than doing that on separate branch, as we will also need to align that with the ongoing changes (unless we're rewriting everything from scratch...). The other case is how to deal with the acceptance tests of the public api for old and new code?

Reply:
I think we end up with almost brand new syntax for defining and configuring projections. I don't think the public API for appending events or querying events would change too much. 

I do think the `IEvent` and whatever else we come with for streams will absolutely change. The underlying table structures are almost 100% certain to have additions

The acceptance tests are still usable, because that's mostly running events through and seeing expected states for the projected documents at the end.

Reply:
We can play that by ear. But my main concern is that we won't end up with something that will be hard to integrate. With git, we at least have built-in conflicts resolution etc. I'd like to also minimize the number of breaking change in the API.

However this is a good point that our projections api is far from being ideal.